### DISCUSSION TITLE
Static Scenes

### DISCUSSION THREAD

**Discussion by nlupugla** [godotengine/godot-proposals#7571] [Engine Core]
It would be great if there was a way to tell the engine that certain aspects of a scene's tree will never change. This could provide many advantages, such as node type safety, better editor autocompletion, and potentially performance improvements.

Consider a simple example of a Character scene laid out as
 - Character (Node2D)
 - - Sprite (Sprite2D)
 - - HealthBar (ProgressBar)
 
It's easy to imagine that you would want all Character instances to have a valid Sprite of type Sprite2D and valid HealthBar of type ProgressBar. Currently, the engine provides no built-in way to enforce such a constraint; there's nothing preventing you from instancing a Character scene, then deleting/renaming/moving/replacing the Sprite or ProgressBar node. This lack of constraint has a few important consequences:
1. `$Sprite` is not cached for fast lookup (see https://github.com/godotengine/godot-proposals/issues/996)
2. `$Sprite` is typed as a `Node`, not `Sprite2D`. 
3. If you inadvertently delete/rename/move/replace `Sprite` somewhere in code, the engine can only warn you about this at runtime when you try to access `$Sprite` and the node isn't found.
This static scene proposal aims to solve all of the above issues.

Here's how this might look in practice. In the editor, there is a way to mark a node as "static". This could be a checkbox near where the "Access as Scene Unique Node" option appears. Checking this option makes that node, and all of its children, "static", meaning they cannot be deleted/renamed/moved/replaced, although they can still acquire new children. A script can then take advantage of this property by declaring the node as static in some way. 

For example

```GDScript
class_name Character 
attaches_to "^Character" in "res://character.tscn"
extends Node2D

func _ready() -> void:
    $Sprite.visible = true
    $ProgressBar.visible = true
    # the lines above got to benefit from autocomplete because the editor knew their types
```
Such a script will only be valid if a "static" node exists in the file "res://character.tsn" at path "^Character". In addition, the editor will highlight as an error any line that attempts to delete/rename/move/replace `Sprite` or `HealthBar`. 

In summary, there are many cases where the engine user knows that certain parts of the scene tree are fixed. This proposal gives the user a way to inform the engine and benefit from that information. I would love to hear thoughts from the community on this idea!

**Comment by nlupugla** [comment]
It was recommended that I move this discussion to an issue. Please see https://github.com/godotengine/godot-proposals/issues/7572 if you would like to share your thoughts.
